src/protocol/payload/reject.rs

```diff
@@ -0,0 +1,93 @@
+use crate::protocol::payload::{read_n_bytes, VarStr};
+
+use std::io::{self, Cursor, Read, Write};
+
+#[derive(Debug)]
+pub struct Reject {
+    message: VarStr,
+    ccode: CCode,
+    reason: VarStr,
+    data: Vec<u8>,
+}
+
+impl Reject {
+    pub fn encode(&self, buffer: &mut Vec<u8>) -> io::Result<()> {
+        self.message.encode(buffer)?;
+        self.ccode.encode(buffer)?;
+        self.reason.encode(buffer)?;
+        buffer.write_all(&self.data)
+    }
+
+    pub fn decode(bytes: &mut Cursor<&[u8]>) -> io::Result<Self> {
+        let message = VarStr::decode(bytes)?;
+        let ccode = CCode::decode(bytes)?;
+        let reason = VarStr::decode(bytes)?;
+
+        let mut data = Vec::new();
+        bytes.read_to_end(&mut data)?;
```

Comment:
As the length of that field is implied in the body length, reading into a vec with `read_to_end` is probably fine. 

Note: another option would have been to make it a fixed length of 32 bytes ([zebra seems to make that assumption](https://github.com/ZcashFoundation/zebra/blob/24bf952e982bde28eb384b211659159d46150f63/zebra-network/src/protocol/external/codec.rs#L508-L515)).

Reply:
`Optional extra data provided by some errors. Currently, all errors which provide this field fill it with the TXID or block header hash of the object being rejected, so the field is 32 bytes`

so current behaviour is `Option<[u8; 32]>` -- not sure which we prefer?